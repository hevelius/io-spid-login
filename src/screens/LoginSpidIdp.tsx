import React, {Component} from 'react';
import {WebView} from 'react-native-webview';
import {WebViewNavigation} from 'react-native-webview';
import {onLoginUriChanged} from '../utils/login';
import {SessionToken} from '../utils/SessionToken';

interface Props {
  profile_url: string;
  onLoginSuccess: (token: SessionToken) => void;
  onLoginFailure: (error: Error) => void;
}

type State = {
  errorCode?: string;
  loginTrace?: string;
};

export class LoginSpidIdp extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      loginTrace: this.props.profile_url,
    };
  }

  private handleShouldStartLoading = (event: WebViewNavigation): boolean => {
    const isLoginUrlWithToken = onLoginUriChanged(
      this.handleLoginFailure,
      this.handleLoginSuccess,
    )(event);
    // URL can be loaded if it's not the login URL containing the session token - this avoids
    // making a (useless) GET request with the session in the URL
    return !isLoginUrlWithToken;
  };

  private handleLoadingError = (): void => {
    console.log('handleLoadingError');
  };

  private handleLoginFailure = (errorCode?: string) => {
    this.props.onLoginFailure(
      new Error(`login failure with code ${errorCode || 'n/a'}`),
    );
  };

  private handleLoginSuccess = (token: SessionToken) => {
    this.props.onLoginSuccess(token);
  };

  render() {
    return (
      <WebView
        source={{uri: this.props.profile_url}}
        textZoom={100}
        onError={this.handleLoadingError}
        javaScriptEnabled={true}
        onShouldStartLoadWithRequest={this.handleShouldStartLoading}
      />
    );
  }
}
