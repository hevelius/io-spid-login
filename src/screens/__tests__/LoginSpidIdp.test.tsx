import React from 'react';
import renderer from 'react-test-renderer';

import {LoginSpidIdp} from '../LoginSpidIdp';

it('renders correctly with defaults', () => {
  const webview = renderer
    .create(
      <LoginSpidIdp
        profile_url={'https://github.com'}
        onLoginSuccess={(_: any) => _}
        onLoginFailure={(_: Error) => _}
      />,
    )
    .toJSON();
  expect(webview).toMatchSnapshot();
});
