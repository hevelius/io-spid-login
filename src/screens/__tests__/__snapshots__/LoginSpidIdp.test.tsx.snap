// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with defaults 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <RNCWebView
    cacheEnabled={true}
    injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
    injectedJavaScriptForMainFrameOnly={true}
    javaScriptEnabled={true}
    messagingEnabled={false}
    onContentProcessDidTerminate={[Function]}
    onError={[Function]}
    onHttpError={[Function]}
    onLoadingError={[Function]}
    onLoadingFinish={[Function]}
    onLoadingProgress={[Function]}
    onLoadingStart={[Function]}
    onMessage={[Function]}
    onShouldStartLoadWithRequest={[Function]}
    source={
      Object {
        "uri": "https://github.com",
      }
    }
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        Object {
          "backgroundColor": "#ffffff",
        },
        undefined,
      ]
    }
    textZoom={100}
    useSharedProcessPool={true}
  />
</View>
`;
